micrometer-core/src/main/java/io/micrometer/core/instrument/binder/kafka/KafkaMetrics.java

```diff
@@ -0,0 +1,279 @@
+    }
+
+    private void checkAndRegisterMetrics(MeterRegistry registry) {
+        Map<MetricName, ? extends Metric> metrics = metricsSupplier.get();
+        if (currentSize.get() != metrics.size()) { // only triggered when number of metrics change
+            currentSize.set(metrics.size());
+            Map<String, Set<Meter>> registered = new HashMap<>();
+            //TODO filter out the following metrics: count (num of metrics), app.info metadata
+            // @jeqo: should we add app.info fields as tags? e.g. version and commit.id could be useful
+            metrics.forEach((name, metric) -> {
+                // register metric
+                String metricName = metricName(metric);
+                Meter meter;
+                if (metricName.endsWith("total")
+                        || metricName.endsWith("count")) {
+                    meter = registerCounter(registry, metric, metricName, extraTags);
+                } else if (metricName.endsWith("min")
+                        || metricName.endsWith("max")
+                        || metricName.endsWith("avg")) {
+                    meter = registerGauge(registry, metric, metricName, extraTags);
+                } else if (metricName.endsWith("rate")) {
+                    meter = registerTimeGauge(registry, metric, metricName, extraTags);
+                } else { // this filter might need to be more extensive.
+                    meter = registerGauge(registry, metric, metricName, extraTags);
+                }
+                // collect metrics with same name to validate number of labels
+                Set<Meter> meters = registered.get(metricName);
+                if (meters == null) meters = new HashSet<>();
+                meters.add(meter);
+                registered.put(metricName, meters);
+            });
+
+            // clean up metrics with low number of tags
+            registered.forEach((metricName, meters) -> {
+                if (meters.size() > 1) {
+                    // find largest number of tags
+                    int maxTagsSize = 0;
+                    for (Meter meter : meters) {
+                        int size = meter.getId().getTags().size();
+                        if (maxTagsSize < size) maxTagsSize = size;
+                    }
+                    // remove meters with lower number of tags
+                    for (Meter meter : meters) {
+                        if (meter.getId().getTags().size() < maxTagsSize) registry.remove(meter);
+                    }
+                }
+            });
+        }
+    }
+
+    private TimeGauge registerTimeGauge(MeterRegistry registry, Metric metric, String metricName,
+                                        Iterable<Tag> extraTags) {
+        return TimeGauge.builder(
+                metricName, metric, TimeUnit.SECONDS, m -> {
+                    checkAndRegisterMetrics(registry);
+                    if (m.metricValue() instanceof Double) {
```

Comment:
Minor but I find this if-statement more readable when extracted and reused

Reply:
abstracting this in a single method.